Improve serialization of non-finite values

Meorawr opened the pull request:
Serializing non-finite values such as NaNs and infinity would cause an error as the check for fractional values in the `number` type writer wouldn't consider such inputs, causing them to instead be written out as single-byte integers and causing errors in later string.char calls. The following test cases would manifest the error shown below:

```lua
LibSerialize:Serialize(0/0) -- This one would work outside of the WoW environment, but not inside.
LibSerialize:Serialize(math.huge)
LibSerialize:Serialize(-math.huge)
```

```
1x integer overflow attempting to store -nan(ind)
[string "=[C]"]: ?
[string "@LibSerialize\LibSerialize-1.lua"]:541: in function <LibSerialize\LibSerialize.lua:539>
[string "@LibSerialize\LibSerialize-1.lua"]:902: in function `_WriteInt'
[string "@LibSerialize\LibSerialize-1.lua"]:898: in function `_WriteByte'
[string "@LibSerialize\LibSerialize-1.lua"]:974: in function `writeFn'
[string "@LibSerialize\LibSerialize-1.lua"]:893: in function `_WriteObject'
[string "@LibSerialize\LibSerialize-1.lua"]:1248: in function <LibSerialize\LibSerialize.lua:1229>
```

This change improves the handling of non-finite values so that, at the very least, they don't error in the serialization/deserialization process and are now actively tested.

The following cases now correctly serialize and deserialize to their original values without errors:

  - Infinity
  - -Infinity
  - NaN

The following cases have improved such that they won't error upon serialization, but won't round trip to their original values:

  - QNaN (will deserialize as NaN)

A few quirks have been observed due to the WoW environment being compiled with floating point optimizations (`/fp:fast`) compared to a typical Lua 5.1 interpreter that isn't.

The first quirk is that NaN values are inequal to themselves outside the game, but inside the WoW environment all comparisons involving NaNs are actually true.

The next quirk is that the stringified forms of various non-finite values differ between the two environments, which would have led to a few subtle serialization differences - one example is that the string representation of infinity in WoW is `"inf"` and can be converted back from a string via tonumber, whereas elsewhere it'll be `"#1.INF"` and *can't* roundtrip through tostring/tonumber.

In terms of what's not implemented, the main one is support for quiet NaNs (qNaNs) as generated by the expression `-(0/0)`. Handling of these is improved in the sense that they no longer error, but they won't round trip though the library and will deserialize as standard NaNs. These should be rare enough to not matter for now, and having them not error is probably good enough.

Signalling NaNs (sNaNs) are untested as at present I'm unsure if it's possible to get a representation of them in Lua whatsoever.

Finally, serializing tables where the key is a NaN is left unsupported entirely. Outside of the WoW environment such a key is actually invalid and will error upon insertion, whereas ingame the insertion works but serialization seems to hard crash the client. I feel supporting this particular case is _probably_ out of scope given the stock VM has an explicit error condition in place and it's not likely Blizzard actively intended for this to be supported ingame.

Meorawr committed:
Resolve issues with serializing NaN values

Previously, serializing NaN values (0/0) would cause an error as the
check for fractional values wouldn't consider NaNs as needing to be
encoded as floating point values, causing them to instead be written out
as single-byte integers and causing errors in later string.char calls.

This change improves the handling of NaNs and non-finite values so that,
at the very least, they don't error in the serialization/deserialization
process and are now actively tested.

The following cases now correctly serialize and deserialize without
errors:

  - Infinity
  - -Infinity
  - NaN

The following cases have improved such that they won't error upon
serialization, but won't round trip to their original values:

  - QNaN (will deserialize as NaN)

A few quirks have been observed however due to the WoW environment
being compiled with floating point optimizations (/fp:fast) compared to
a typical Lua 5.1 interpreter that isn't.

The first quirk is that NaN values are inequal to themselves outside the
game, but inside the WoW environment all comparisons involving NaNs are
actually true.

The next quirk is that the stringified forms of various non-finite values
differ between the two environments, which would have led to a few subtle
serialization differences - one example is that the string representation
of infinity in WoW is "inf" and can be converted to/from a string via
tonumber, whereas elsewhere it'll be "#1.INF" and *can't* roundtrip
through tostring/tonumber.

In terms of what's not implemented, the main one is support for quiet
NaNs (QNaNs) as generated by the expression "-(0/0)". Handling of these
is improved in the sense that they no longer error, but they won't round
trip though the library and will deserialize as standard NaNs. These
should be rare enough to not matter for now, and having them not error
is probably good enough.

Signalling NaNs (SNaNs) are untested as at present I'm unsure if it's
possible to get a representation of them in Lua whatsoever.

Finally, serializing tables where the key is a NaN is left unsupported
entirely. Outside of the WoW environment such a key is actually invalid,
whereas ingame it works but actually seems to hard crash the client upon
serialization.

Meorawr committed:
Add test case for serializing -0

Just as a peace of mind thing, since negative zero has annoying NaN-like
properties ingame such as comparing equal to positive zero.

Meorawr committed:
Bump minor version

Meorawr committed:
Slightly restructure FloatToString

Check for NaN inputs first to save on a call to frexp, and clarify why
we explicitly test for NaN values when converting infinity values.

Meorawr commented on LibSerialize.lua:
```diff
@@ -958,7 +983,7 @@ LibSerialize._WriterTable = {
             end
             local asString = tostring(numAbs)
-            if #asString < 7 and tonumber(asString) == numAbs then
+            if #asString < 7 and tonumber(asString) == numAbs and IsFinite(numAbs) then
```
Just as a pre-emptive clarification, the IsFinite test here is to handle an edge case where `math.huge` is serialized ingame, exported via saved variables, and then deserialized out of game. The reasoning being `tonumber("inf")` round trips back to `numAbs` ingame but not outside, so by explicitly disallowing non-finite values we force them to consistently serialize as NUM_FLOAT types at the expense of such values being a few bytes larger.

rossnichols replied:
Pre-emptive clarifications sound like a good place for comments :)

Also, do you see a clean way to avoid the tostring/tonumber round-trip for this case? Technically wasted effort if the number is `inf`.

Meorawr committed:
Change IsNaN test to be a bit safer

It should be impossible for a valid non-NaN number to be simultaneously
lesser than and greater-or-equal to zero, and this seems to hold up
reliably both ingame and out of game.

Meorawr committed:
Fix IsFinite test under -ffast-math

GCC's -ffast-math optimizations seem to have additional quirks with NaN
values causing the IsFinite test to fail as NaNs in such an environment
will compare equally to both +/- infinity but not <= or >=, somehow.

Meorawr committed:
Simplify IsFinite test

Rather than special case it for optimization we'll just use the original
logic and defer the IsNaN test out to the actual function dedicated for
it.

rossnichols commented on LibSerialize.lua:
```diff
@@ -1082,7 +1105,7 @@ LibSerialize._WriterTable = {
             local entireMapSerializable = true
             for k, v in pairs(tab) do
-                local isArrayKey = type(k) == "number" and k >= 1 and k <= arrayCount and not IsFractional(k)
+                local isArrayKey = type(k) == "number" and k >= 1 and k <= arrayCount and not IsFloatingPoint(k)
```
Mind refactoring this into a small helper function, since it's repeated three times? You didn't introduce it, but would be nice to clean up.

rossnichols commented:
It looks like you added support for running the tests within WoW too, with the LibStub change. I think it makes sense to follow the pattern of other addons and add a TOC file so it can be loaded as an addon. Perhaps wrapping the tests inside e.g. `LibSerialize:RunTests()` if loaded in this environment, but `lua tests.lua` still runs immediately?

I can follow up this PR with the machinery needed for this, but if you already have it locally feel free to attach. For a project as small as this, I'm not concerned about the PR doing too many things :)

Meorawr committed:
Add IsArrayKey function

Meorawr committed:
Add note on IsFinite check for number encoding

Meorawr committed:
Add TOC file, defer tests when loaded ingame

Tests will run immediately when the script is loaded from a non-WoW
environment, however in-game will require a manual call to
LibSerialize:RunTests().
